lib/scholar/neighbors/random_projection_forest.ex

```diff
@@ -0,0 +1,363 @@
+defmodule Scholar.Neighbors.RandomProjectionForest do
+  @moduledoc """
+  Random Projection Forest.
+
+  Each tree in a forest is constructed using a divide and conquer approach.
+  We start with the entire dataset and at every node we project the data onto a random
+  hyperplane and split it in the following way: the points with the projection smaller
+  than or equal to the median are put into the left subtree and the points with projection
+  greater than the median are put into the right subtree. We then proceed
+  recursively with the left and right subtree.
+  In this implementation the trees are complete, i.e. there are 2^l nodes at level l.
+  The leaves of the trees are arranged as blocks in the field `indices`. We use the same
+  hyperplane for all nodes on the same level as in [2].
+
+  * [1] - Random projection trees and low dimensional manifolds
+  * [2] - Fast Nearest Neighbor Search through Sparse Random Projections and Voting
+  """
+
+  import Nx.Defn
+  import Scholar.Shared
+  require Nx
+
+  @derive {Nx.Container,
+           keep: [:depth, :leaf_size, :num_trees],
+           containers: [:indices, :data, :hyperplanes, :medians]}
+  @enforce_keys [:depth, :leaf_size, :num_trees, :indices, :data, :hyperplanes, :medians]
+  defstruct [:depth, :leaf_size, :num_trees, :indices, :data, :hyperplanes, :medians]
+
+  opts = [
+    num_trees: [
+      required: true,
+      type: :pos_integer,
+      doc: "The number of trees in the forest."
+    ],
+    min_leaf_size: [
+      required: true,
+      type: :pos_integer,
+      doc: "The minumum number of points in the leaf."
+    ],
+    key: [
+      type: {:custom, Scholar.Options, :key, []},
+      doc: """
+      Used for random number generation in hyperplane initialization.
+      If the key is not provided, it is set to `Nx.Random.key(System.system_time())`.
+      """
+    ]
+  ]
+
+  @opts_schema NimbleOptions.new!(opts)
+
+  @doc """
+  Grows a random projection forest.
+
+  ## Options
+
+  #{NimbleOptions.docs(@opts_schema)}
+
+  ## Examples
+
+      iex> key = Nx.Random.key(12)
+      iex> tensor = Nx.iota({5, 2})
+      iex> Scholar.Neighbors.RandomProjectionForest.fit(tensor, num_trees: 3, min_leaf_size: 2, key: key).indices
```

Comment:
Instead of returning only the indices, you can assign the forest to a variable and then assert on multiple entries?

```suggestion
      iex> forest = Scholar.Neighbors.RandomProjectionForest.fit(tensor, num_trees: 3, min_leaf_size: 2, key: key)
      iex> forest.indices
```

Things like depth, data, and so on would be nice to show.